catalog/core/catalog-core-migratable/src/main/java/org/codice/ddf/catalog/migratable/impl/CatalogMigratableImpl.java

```diff
@@ -0,0 +1,221 @@
+     * @param config        export configuration information
+     */
+    public CatalogMigratableImpl(@NotNull String description, @NotNull CatalogFramework framework,
+            @NotNull FilterBuilder filterBuilder, @NotNull MigrationFileWriter fileWriter,
+            @NotNull CatalogMigratableConfig config) {
+        super(description, true);
+
+        notNull(framework, "CatalogFramework cannot be null");
+        notNull(filterBuilder, "FilterBuilder cannot be null");
+        notNull(fileWriter, "FileWriter cannot be null");
+        notNull(config, "Configuration object cannot be null");
+
+        this.framework = framework;
+        this.filterBuilder = filterBuilder;
+        this.fileWriter = fileWriter;
+        this.config = config;
+    }
+
+    /**
+     * Exports all the metacards currently stored in the catalog framework.
+     * <p>
+     * {@inheritDoc}
+     */
+    public MigrationMetadata export(Path exportPath) throws MigrationException {
+        taskManager = createTaskManager(config, createExecutorService(config));
+        config.setExportPath(exportPath.resolve(CATALOG_EXPORT_DIRECTORY));
+        fileWriter.createExportDirectory(config.getExportPath());
+
+        Collection<MigrationWarning> warnings = new ArrayList<>();
+        Map<String, Serializable> props = createMapWithNativeQueryMode();
+        Filter dumpFilter = filterBuilder.attribute(Metacard.ANY_TEXT)
+                .is()
+                .like()
+                .text("*");
+
+        QueryImpl exportQuery = createQuery(dumpFilter);
+        exportQuery.setPageSize(config.getExportQueryPageSize());
+        exportQuery.setRequestsTotalResultsCount(false);
+
+        QueryRequest exportQueryRequest = new QueryRequestImpl(exportQuery, props);
+
+        Subject subject = retrieveSystemSubject();
+        if (subject == null) {
+            LOGGER.error("System Subject was null");
+            throw new ExportMigrationException(DEFAULT_FAILURE_MESSAGE);
+        }
+
+        try {
+            subject.execute(new CatalogMigrationRunner(exportQuery, exportQueryRequest));
+        } catch (RuntimeException e) {
+            LOGGER.error("Internal error occurred when exporting catalog: {}", e.getMessage(), e);
+            throw new ExportMigrationException(DEFAULT_FAILURE_MESSAGE);
+        } finally {
+            taskManager.exportFinish();
+        }
+
+        return new MigrationMetadata(warnings);
+    }
+
+    // Factory method for creating QueryImpl objects so they can be mocked later.
+    QueryImpl createQuery(Filter dumpFilter) {
+        return new QueryImpl(dumpFilter);
+    }
+
+    // Factory method for creating MigrationTaskManager objects so they can be mocked later.
+    MigrationTaskManager createTaskManager(CatalogMigratableConfig config,
+            ExecutorService executorService) {
+        return new MigrationTaskManager(config, fileWriter, executorService);
```

Comment:
I agree that we should reduce the number of places where the factory method pattern is used for the sole purpose of unit testing. However, there really are legitimate places where this can be used, like this one. Using a real object here would make unit testing this code real difficult. Which means that we're pretty much left with the lesser of three evils: factory object that can be injected and mocked, factory method, or PowerMock. I really think that in this case, the factory method was the right option.